File: SeaCleaner/Client/Game/Sprite.cs
using Microsoft.JSInterop;
using System;
using System.Threading.Tasks;

namespace SeaCleaner.Client.Game
{
    internal class Sprite
    {
        public double PosX { get; set; }
        public double PosY { get; set; }
        public double DisplayWidth { get; private set; }
        public double DisplayHeight { get; private set; }
        public SpriteImageInfo Image { get; private set; }
        public Sprite(double x, double y, double w, double h, SpriteImageInfo image)
        {
            PosX = x;
            PosY = y;
            DisplayWidth = w;
            DisplayHeight = h;
            Image = image;
        }

        public virtual async ValueTask Draw(IJSRuntime jsRuntime)
        {
            await jsRuntime.InvokeVoidAsync("drawImage", Image.SpriteName, PosX, PosY, DisplayWidth, DisplayHeight);
        }
    }

    internal class Animation
    {
        public double PosX { get; protected set; }
        public double PosY { get; protected set; }
        public SpriteImageInfo Image { get; private set; }
        public int UpdatesOnFrame { get; private set; }
        public double UpdatesShiftX { get; protected set; } 
        public double UpdatesShiftY { get; protected set; }
        private int _updatesCounter = 0;
        private int _currentFrame = 0;
        protected Random Rand { get; private set; } = new Random();

        public Animation(double x, double y, int updates, double shiftX, double shiftY, SpriteImageInfo image)
        {
            PosX = x;
            PosY = y;
            UpdatesOnFrame = updates;
            UpdatesShiftX = shiftX;
            UpdatesShiftY = shiftY;
            Image = image;
        }

        public virtual void Update()
        {
            PosX += UpdatesShiftX;
            PosY += UpdatesShiftY;

            _updatesCounter++;
            if (_updatesCounter >= UpdatesOnFrame)
            {
                if (Image.FramesCount <= _currentFrame + 1)
                    _currentFrame = 0;
                else
                    _currentFrame++;

                _updatesCounter = 0;
            }
        }

        private int framePosX;
        private int framePosY;

        public virtual async ValueTask Draw(IJSRuntime jsRuntime)
        {
            framePosX = Image.Vertical ? 0 : _currentFrame * Image.FrameWidth;
            framePosY = Image.Vertical ? _currentFrame * Image.FrameHeight : 0;

            await jsRuntime.InvokeVoidAsync("drawSprite", Image.SpriteName, framePosX, framePosY, Image.FrameWidth, Image.FrameHeight, PosX, PosY);
        }
    }

    internal class Bubble : Animation
    {
        public double Fluctuation { get; private set; }  
        private readonly double _fluctCoeff = 0.1;
        public Bubble(double x, double y, int updates, double shiftY, double fluctuation, SpriteImageInfo imageInfo) : base(x, y, updates, 0, shiftY, imageInfo)
        {
            Fluctuation = fluctuation;
        }

        public override void Update()
        {
            PosX += Math.Sin(PosY * _fluctCoeff) * Fluctuation;

            if (PosY <= 150)
            {
                PosY = 720 + Math.Round(Rand.NextDouble() * 1000) % 30;
                PosX = Rand.NextDouble() * Game.SCREEN_WIDTH;
            }
            base.Update();
        }
    }

    internal class Fishes : Animation
    {
        public bool ToLeft { get; private set; }
        public Fishes(int updates, double shiftX, bool toLeft, SpriteImageInfo imageInfo): base (300, 200, updates, shiftX, 0, imageInfo)
        {
            ToLeft = toLeft;            
        }

        public void Initialize()
        {            
            UpdatesShiftX = Math.Round(Rand.NextDouble() * 1000) % 4 + 3;
            PosY = Rand.NextDouble() * 1000 % 480 + 170;

            if (ToLeft)
            {
                PosX = 1050 + Math.Round(Rand.NextDouble() * 1000 / 3);
                if (UpdatesShiftX > 0) UpdatesShiftX = -UpdatesShiftX;
            }
            else
            {
                PosX = -50 - Math.Round(Rand.NextDouble() * 1000 / 3);
                if (UpdatesShiftX < 0) UpdatesShiftX = -UpdatesShiftX;
            }
        }

        public bool IsOffScreen() => (ToLeft, PosX < -50, PosX > 1050) switch
        {
            (true, true, _) => true,
            (true, false, _) => false,
            (false, _, true) => true,
            (false, _, false) => false
        };
    }
}

File: SeaCleaner/Client/wwwroot/js/Sprite.js
//
// Класс для хранения неанимированного изображения 
// _X, _Y - координаты изображения
// _H, _W - высота и ширина картинки при ее отрисовке
function CSprite(_X, _Y, _W, _H){
    this.PosX = _X;
    this.PosY = _Y;
    this.Width = _W;
    this.Height = _H;
    // метод загрузки картинки   
    this.loadSprite = function(ImageFileName){
        this.pic = new Image();
        this.pic.src = ImageFileName;
    };
    // метод для вывода картинки на канву
    this.draw = function(context){
        context.drawImage( 
            this.pic, // собственно, картинка
            this.PosX, this.PosY,  // позиция вывода
            this.Width, this.Height); //  и размеры
        
    };    
};
/*
 *    класс-анимация использующая картинку из хранилища
 */
function CAnimation(_X, _Y, _Updates, _Shift_X, _Shift_Y, _SpriteName){
    this.PosX = _X;
    this.PosY = _Y;
    this.UpdatesOnFrame = _Updates;
    this.UpdatesShiftX = _Shift_X; // 
    this.UpdatesShiftY = _Shift_Y;
    this.SpriteName = _SpriteName;
    //
    this.UpdatesCounter = 0;
    this.CurrentFrame = 0;
    //
    this.update = function(){
        //
        this.PosX += this.UpdatesShiftX;
        this.PosY += this.UpdatesShiftY;
        //
        if(++this.UpdatesCounter >= this.UpdatesOnFrame){
            if(MultiImage[this.SpriteName].SpriteCount <= this.CurrentFrame + 1)
                this.CurrentFrame = 0;
            else
                this.CurrentFrame++;
            this.UpdatesCounter = 0;
        }
    };
    //
    this.draw = function(context){
        if(MultiImage[this.SpriteName].IsVertical)
            context.drawImage(
                MultiImage[this.SpriteName],
                0, 
                this.CurrentFrame * MultiImage[this.SpriteName].SpriteHeight,
                MultiImage[this.SpriteName].SpriteWidth,
                MultiImage[this.SpriteName].SpriteHeight,
                this.PosX,
                this.PosY,
                MultiImage[this.SpriteName].SpriteWidth,
                MultiImage[this.SpriteName].SpriteHeight         
            );
        else
            context.drawImage(
                MultiImage[this.SpriteName],
                this.CurrentFrame * MultiImage[this.SpriteName].SpriteWidth, 
                0,
                MultiImage[this.SpriteName].SpriteWidth,
                MultiImage[this.SpriteName].SpriteHeight,
                this.PosX,
                this.PosY,
                MultiImage[this.SpriteName].SpriteWidth,
                MultiImage[this.SpriteName].SpriteHeight         
            );                    
    };
};
/* 
 *  Класс Пузырь.
 */
function CBubble(_X, _Y, _Updates, _Shift_Y, _Fluctuation){
    CAnimation.apply(this, [_X, _Y, _Updates, 0, _Shift_Y, "bubble"]);
    this._Update = this.update;
    this.fluctuation = _Fluctuation; // повышает синусоиду 
    this.fluctCoeff = 0.1; // удлиняет синусоиду
    //
    this.update = function(){
        // движение пузыря по синусоиде
        this.PosX += Math.sin(this.PosY * this.fluctCoeff)*this.fluctuation;
        // автоматическое перемещение на дно в случае достижения волн
        if(this.PosY <= 150){
            this.PosY = 720 + Math.round(Math.random()*1000) % 30;
            this.PosX = Math.random() * SCREEN_WIDTH; // получается как раз 0 <= X <= 1000
        }                     
        //
        this._Update();        
    };
};
/*
 * Класс Стая рыбок
 */
function CFishes(_Updates, _Shift_X, _toLeft, _spriteName){
    CAnimation.apply(this, [300, 200, _Updates, _Shift_X, 0, _spriteName]);
    this.toLeft = _toLeft;
    //
    this.initialize = function(){
        this.UpdatesShiftX = Math.round(Math.random()*1000) % 4 + 3;
        this.PosY = Math.random() * 1000 % 480/*550*/ + 170; 
        if(this.toLeft){
            this.PosX = 1050 + Math.round(Math.random() * 1000 / 3);
            if(this.UpdatesShiftX > 0)
                this.UpdatesShiftX = -this.UpdatesShiftX;
        }
        else{
            this.PosX = -50 - Math.round(Math.random() * 1000 / 3);
            if(this.UpdatesShiftX < 0)
                this.UpdatesShiftX = -this.UpdatesShiftX;
        }       
    };
    this.isOffScreen = function(){
        if(this.toLeft)
            if(this.PosX < -50)
                return true;
            else
                return false;
        else
            if(this.PosX > 1050)
                return true;
            else
                return false;
    };
};
